src/Features/CSharp/Portable/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // We're going to be checking every invocation in the compilation. Cache information
+                // we compute in this object so we don't have to continually recompute it.
+                var infoCache = new InfoCache(compilationContext.Compilation);
+                compilationContext.RegisterOperationAction(
+                    c => AnalyzeInvocation(c, infoCache),
+                    OperationKind.Invocation);
+            });
+        }
+
+        private void AnalyzeInvocation(
+            OperationAnalysisContext context, InfoCache infoCache)
+        {
+            var cancellationToken = context.CancellationToken;
+            var invocation = (IInvocationOperation)context.Operation;
+
+            var invocationSyntax = invocation.Syntax as InvocationExpressionSyntax;
+            if (invocationSyntax is null ||
+                invocationSyntax.ArgumentList is null)
+            {
+                return;
+            }
+
+            // Check if we're at least on C# 8, and that the user wants these operators.
+            var syntaxTree = invocationSyntax.SyntaxTree;
+            var parseOptions = (CSharpParseOptions)syntaxTree.Options;
+            if (parseOptions.LanguageVersion < LanguageVersion.CSharp8)
+            {
+                return;
+            }
+
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet is null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CSharpCodeStyleOptions.PreferRangeOperator);
+            if (!option.Value)
+            {
+                return;
+            }
+
+
+            // look for `s.Slice(e1, end - e2)`
+            if (invocation.Instance is null ||
+                invocation.Instance.Syntax is null ||
+                invocation.Arguments.Length != 2)
+            {
+                return;
+            }
+
+            // See if the call is to something slice-like.
+            var targetMethod = invocation.TargetMethod;
+
+            // Second arg needs to be a subtraction for: `end - e2`.  Once we've seen that we have
+            // that, try to see if we're calling into some sort of Slice method with a matching
+            // indexer or overload
+            if (!IsSubtraction(invocation.Arguments[1].Value, out var subtraction) ||
+                !infoCache.TryGetMemberInfo(targetMethod, out var memberInfo))
+            {
+                return;
+            }
+
+            // See if we have: (start, end - start).  Specifically where the start operation it the
+            // same as the right side of the subtraction.
+            var startOperation = invocation.Arguments[0].Value;
+
+            if (CSharpSyntaxFactsService.Instance.AreEquivalent(startOperation.Syntax, subtraction.RightOperand.Syntax))
```

Comment:
>CSharpSyntaxFactsService.Instance.AreEquivalent(startOperation.Syntax, subtraction.RightOperand.Syntax) [](start = 16, length = 103)

Is "syntax is equivalent" check sufficient to guarantee the syntax itself has no side effects? It will probably suffice for most realistic cases.

Reply:
It is *not* sufficient for side effects.  However, i've made myself reasonably satisfied that that's just so unlikely that we don't want to care (i've also doced this).  Specifically, say you have this:

`foo.GetLast()[foo.GetLast().Length - 1]`

What are hte chances that you have written the above *and* are actually *depending* side effects of having the expr executed twice?  I make the claim that it's virtually nil.  And that preventing this from working will cut out far too many cases that actually appear int eh wild when people are working quickly and don't want to bother extracting these exprs into locals.  

In practice, if you have the above (remember, you need the length and the subtraction, and both exprs have to match), you are expecting that the exprs are idempotent.  i.e. running them twice produces the same result as running it once.  

This fits the IDE mentality that it's ok to not be perfect semantics preservation if the only thing that changes are semantics for something that is *super* unlikely to exist.  I also mentioned though that if we learn otherwise, we can move this to be more conservative.  However, i'm fine with this as the starting point :)